File: python/graph/200_Number_of_Islands.py
'''
Time: O(m*n)
'''
import collections

class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        if not grid:  # if no grid, islands = 0
            return 0
        
        rows, cols = len(grid), len(grid[0])  # num of rows, num of cols
        visited = set()  # islands have been visited 
        islands = 0  # num of islands

        # 1 island contain how many "1"s
        def bfs(r, c):
            q = collections.deque()
            visited.add((r, c))
            q.append((r, c))

            while q:
                row, col = q.popleft()  # dfs: q.pop()
                directions = [[-1, 0], [1, 0], [0, -1], [0, 1]]  # left, right, above, bottom

                for dr, dc in directions:
                    r, c = row + dr, col + dc
                    if (r in range(rows)) and (c in range(cols)) and (grid[r][c] == "1") and ((r, c) not in visited):
                        q.append((r, c))
                        visited.add((r, c))


        # how many islands
        for r in range(rows): 
            for c in range(cols):
                if grid[r][c] == "1" and (r, c) not in visited:
                    bfs(r, c)
                    islands += 1

        return islands

File: javascript/graph/200_Number_of_Islands.js
/**
 * @param {character[][]} grid
 * @return {number}
 */
// Solution 1
var numIslands = function(grid) {
    if (!grid) return 0;

    const visited = new Set();
    const rows = grid.length, cols = grid[0].length;
    let islands = 0;
  
    function bfs(r, c) {
      const q = [];
      visited.add(`${r},${c}`);  // can't use [r, c] and (r, c)
      q.push([r, c]);
  
      while (q.length) {
        const [row, col] = q.shift();
        const directions = [[-1, 0], [1, 0], [0, -1], [0, 1]];
  
        for (let [dr, dc] of directions) {
          const r = row + dr, c = col + dc;
          // can't use 0 <= r < rows
          if ((0 <= r && r < rows) && (0 <= c && c < cols) && (grid[r][c] == "1") && (!(visited.has(`${r},${c}`)))) {
            visited.add(`${r},${c}`);
            q.push([r, c]);
          }
        }
      }
    }
  
    for (let r = 0; r < rows; r++) {
      for (let c = 0; c < cols; c++) {
        if ((grid[r][c] == "1") && (!(visited.has(`${r},${c}`)))) {
          bfs(r, c);
          islands++;
        }
      }
    }
    return islands;
};

// solution 2
var numIslands = function(grid) {
    if (!grid) return 0;
  
    const rows = grid.length, cols = grid[0].length;
    let islands = 0;
  
    function dfs(r, c) {
      if ((0 <= r && r < rows) && (0 <= c && c < cols) && (grid[r][c] == "1")) {
        grid[r][c] = "0";
        dfs(r - 1, c);
        dfs(r + 1, c);
        dfs(r, c - 1);
        dfs(r, c + 1);
      }
    }
  
    for (let r = 0; r < rows; r++) {
      for (let c = 0; c < cols; c++) {
        if (grid[r][c] == "1") {
          dfs(r, c);
          islands++;
        }
      }
    }
    return islands;
  };